client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -134,56 +131,59 @@ void partialCancel()
-    {
-        Stream<T> stream = Streams.stream(iterator)
-                .flatMap(Streams::stream);
-        if (maxRows > 0) {
-            stream = stream.limit(maxRows);
-        }
-        return stream.iterator();
-    }
-
     @VisibleForTesting
-    static class AsyncIterator<T>
-            extends AbstractIterator<T>
+    static class AsyncResultsPageIterator
+            extends CancellableIterator<List<Object>>
     {
         private static final int MAX_QUEUED_ROWS = 50_000;
         private static final ExecutorService executorService = newCachedThreadPool(
                 new ThreadFactoryBuilder().setNameFormat("Trino JDBC worker-%s").setDaemon(true).build());
 
         private final StatementClient client;
-        private final BlockingQueue<T> rowQueue;
+        private final BlockingQueue<List<Object>> rowQueue;
         // Semaphore to indicate that some data is ready.
         // Each permit represents a row of data (or that the underlying iterator is exhausted).
         private final Semaphore semaphore = new Semaphore(0);
         private final Future<?> future;
+
         private volatile boolean cancelled;
         private volatile boolean finished;
 
-        public AsyncIterator(Iterator<T> dataIterator, StatementClient client)
+        AsyncResultsPageIterator(StatementClient client, Consumer<QueryStats> progressCallback, WarningsManager warningsManager, Optional<BlockingQueue<List<Object>>> queue)
         {
-            this(dataIterator, client, Optional.empty());
-        }
+            requireNonNull(progressCallback, "progressCallback is null");
+            requireNonNull(warningsManager, "warningsManager is null");
 
-        @VisibleForTesting
-        AsyncIterator(Iterator<T> dataIterator, StatementClient client, Optional<BlockingQueue<T>> queue)
-        {
-            requireNonNull(dataIterator, "dataIterator is null");
             this.client = client;
             this.rowQueue = queue.orElseGet(() -> new ArrayBlockingQueue<>(MAX_QUEUED_ROWS));
             this.cancelled = false;
             this.finished = false;
             this.future = executorService.submit(() -> {
                 try {
-                    while (!cancelled && dataIterator.hasNext()) {
-                        rowQueue.put(dataIterator.next());
-                        semaphore.release();
+                    do {
+                        QueryStatusInfo results = client.currentStatusInfo();
+                        progressCallback.accept(QueryStats.create(results.getId(), results.getStats()));
+                        warningsManager.addWarnings(results.getWarnings());
+                        Iterable<List<Object>> data = client.currentData().getData();
+                        if (data != null) {
+                            for (List<Object> row : data) {
+                                rowQueue.put(row);
+                                semaphore.release();
```

Comment:
in finally section you first release semaphore and the set finished flag to true. It used to be like that before. But seems wrong.